rulesets/src/collections.ruleset.yml

```diff
@@ -0,0 +1,136 @@
+    description: Response bodies from collection endpoints must offer paging capability.
+    severity: error
+    given: $.paths[?(!@property.match(/.*\/\{[^}]+\}$/))].get.responses['200'].content.application/json.schema.properties
+    then:
+      - field: "paging"
+        function: truthy
+      - field: "paging.type"
+        function: pattern
+        functionOptions:
+          match: "object"
+
+  ##### Root Element #####
+  sps-collection-missing-results-array:
+    description: Response bodies must have a root element called results and is an array of objects.
+    severity: error
+    given: $.paths[?(!@property.match(/.*\/\{[^}]+\}$/))].get.responses['200'].content.application/json.schema.properties.results
+    then:
+      - field: type
+        function: pattern
+        functionOptions:
+          match: "array"
+      - field: items.type
+        function: pattern
+        functionOptions:
+          match: "object"
+
+  ##### Pagination #####
+  sps-missing-pagination-query-parameters:
+    description: 'Collection GET endpoints must support pagination using query parameters: limit, offset, cursor.'
+    severity: error
+    given: $.paths[?(!@property.match(/.*\/\{[^}]+\}$/))].get
+    then:
+      - field: parameters
+        function: schema
+        functionOptions:
+          schema:
+            type: array
+            items:
+              type: object
+              properties:
+                name:
+                  type: string
+              required:
+                - name
+            minItems: 3
+            contains:
+              anyOf:
+                - properties:
+                    name:
+                      const: offset
+                - properties:
+                    name:
+                      const: limit
+                - properties:
+                    name:
+                      const: cursor
+
+  sps-post-request-body-missing-paging-object:
+    description: "POST collection endpoints MUST have a request body schema that includes paging parameters."
+    severity: error
+    given: $.paths[?(!@property.match(/.*\/\{[^}]+\}$/))].post.requestBody.content.application/json.schema.properties.paging
+    then:
+      field: "type"
+      function: pattern
+      functionOptions:
+        match: "object"
+
+  ##### FILTERING #####
+  sps-disallow-resource-identifier-filtering:
+    description: "Resource identifier filtering is not allowed. Use the resource identifier in the URL path."
+    severity: warn
+    given: $.paths..get.parameters.[?(@.in=='query' && @.name=='id')]
+    then:
+      field: "name"
+      function: pattern
+      functionOptions:
+        notMatch: "^id$"
+
+  sps-unreasonable-query-parameters-limit:
+    description: "Filtering query parameters SHOULD have a reasonable limit, no more than 12."
+    severity: warn
+    given: $.paths..get
+    then:
+      function: schema
+      functionOptions:
+        dialect: "draft2020-12"
```

Comment:
why not "auto" for this?

Reply:
When set to auto or the field is absent a unit testing fail but after diving into this more today. I believe this rule may not be possible. The goal of this rule is have a limit on how many query parameters can be present on a given GET endpoint. When asking github copilot for help, it came back with this:

"For the purpose of JSON Schema validation, you can ensure that the array does not exceed a certain size overall, but applying a conditional count based on item properties is beyond the capabilities of JSON Schema as of the latest drafts (up to draft 2020-12)."

This leads me to believe that when this dialect field is persent that unit test is returning a false postive, like erroring on something else than the size limit. When dialect is not present spectral will default it. I am not sure what the default version is but it seems like its at least 2020-12 JSON schema version.

This rule may not be possible with spectral out of the box since we need to count only the query parameters. And the parameters array with OAS can have the following values for its `in` field: `['path', 'query', 'header', 'cookie']`. Just doing a length check is not enough and doing a filter expression within JSONPath does not net us the results we want.

I am comment this rule out and linking to this comment

Reply:
or you just increase it to a larger number and catch overall length without filtering as a best practice - less useful, but still a good boundary. I'm not opposed to remove it as well if you deem value is to low.

Reply:
I commented it out for the time being and we can come back to the rule when we start exploring using JS functions. I think this one would be a fantastic candidate